### PR TITLE
User tests

### PR DESCRIPTION
**Pull request by RCAVelez**
Testing: 
CreateUser
ReadUser Email
ReadUser TripId
AddTripId

### CODE REVIEW

File: frontend/src/TripUtils.js [hunk 1]
```diff
@@ -2,26 +2,22 @@ import { fetchJson } from './fetchJson.js';
 
 export const createTrip = async (email, tripData) => {
   const url = `/api/v1/createTrip?email=${email}&tripData=${JSON.stringify(tripData)}`;
-  const json = await fetchJson(url);
-  return json;
+  await fetch(url, { method: 'POST' });
```
Comment (genedwards) [inline frontend/src/TripUtils.js:5]: you used to return the json response, do you need to do the same?
Updating `fetchJson` to take a 2nd argument might be a good idea
```JS
export const fetchJson = async (url, options = {}) {
  const response = await fetch(url, options);
```

File: backend/src/test/java/com/google/sps/UserCrudTest.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class UserCrudTest {
+  private final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
```
Comment (Cindyzhang977) [inline backend/src/test/java/com/google/sps/UserCrudTest.java:32]: should this also be static?

File: backend/pom.xml [hunk 3]
```diff
@@ -65,20 +65,44 @@
...
+     </dependency>
+     <dependency>
+         <groupId>com.google.appengine</groupId>
+         <artifactId>appengine-tools-sdk</artifactId>
+         <version>1.9.60</version>
+         <scope>test</scope>
+     </dependency>
   </dependencies>
-
+
```
[the remark this answers is not in the document]
Comment (RCAVelez) [reply backend/pom.xml:93]: done

File: backend/src/main/java/com/google/sps/UserCrud.java [hunk 4]
```diff
@@ -28,7 +27,7 @@
   public static Entity createUser(String email) {
     Entity userEntity = new Entity("User");
     userEntity.setProperty("email", email);
-    userEntity.setProperty("tripIds", Collections.emptyList());
+    userEntity.setProperty("tripIds", new ArrayList<String>());
```
[the remark this answers is not in the document]
Comment (RCAVelez) [reply backend/src/main/java/com/google/sps/UserCrud.java:30]: This was a small change that needed to be made.

File: backend/src/main/java/com/google/sps/servlets/createTripServlet.java [hunk 5]
```diff
@@ -15,7 +15,6 @@
 package com.google.sps.servlets;
 
 import com.google.appengine.api.datastore.Entity;
-import com.google.gson.JsonObject;
 import com.google.sps.TripCRUD;
 import com.google.sps.UserCrud;
```
[the remark this answers is not in the document]
Comment (RCAVelez) [reply backend/src/main/java/com/google/sps/servlets/createTripServlet.java:19]: Yes some slight wording mistakes. I will change in a following PR

File: backend/src/main/java/com/google/sps/servlets/createTripServlet.java [hunk 6]
```diff
@@ -45,12 +44,6 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
     }
 
-    Entity tripEntity = TripCRUD.createTrip(email, tripData);
-
-    JsonObject jsonResults = new JsonObject();
-    jsonResults.addProperty("tripId", tripEntity.getKey().getId());
-    jsonResults.addProperty("tripData", tripData);
-    response.setContentType("application/json;");
-    response.getWriter().println(jsonResults);
+    TripCRUD.createTrip(email, tripData);
```
[the remark this answers is not in the document]
Comment (RCAVelez) [reply backend/src/main/java/com/google/sps/servlets/createTripServlet.java:47]: We debated whether it was necessary or not. We decided to not return anything

File: backend/src/main/java/com/google/sps/servlets/createUserServlet.java [hunk 7]
```diff
@@ -29,22 +25,12 @@
...
-    ArrayList<String> tripIds = (ArrayList<String>) userEntity.getProperty("tripIds");
-
-    JsonObject jsonResults = new JsonObject();
-    Gson gson = new Gson();
-    jsonResults.addProperty("email", userEntity.getProperty("email").toString());
-    jsonResults.addProperty("tripIds", gson.toJson(tripIds));
-
-    response.setContentType("application/json;");
-    response.getWriter().println(jsonResults);
+    UserCrud.createUser(email);
```
[the remark this answers is not in the document]
Comment (RCAVelez) [reply backend/src/main/java/com/google/sps/servlets/createUserServlet.java:34]: Same thing. We decided to not return anything

File: backend/src/test/java/com/google/sps/UserCrudTest.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+  private static final String Email = "testEmail@gmail.com";
```
[the remark this answers is not in the document]
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/UserCrudTest.java:33]: ah so the "final" keyword determines a constant. Done

File: backend/src/test/java/com/google/sps/UserCrudTest.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  private static final String Email2 = "testEmail2@gmail.com";
+  private static final Long TripId = Long.parseLong("111222333");
```
[the remark this answers is not in the document]
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/UserCrudTest.java:35]: brain fart

[the remark this answers is not in the document]
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/UserCrudTest.java:35]: done

File: backend/src/test/java/com/google/sps/UserCrudTest.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+  private static final ArrayList<String> EmptyTripIds = new ArrayList<String>();
+  private static final ArrayList<String> SingleTripId = new ArrayList<String>() {
+    { add("111222333"); }
```
[the remark this answers is not in the document]
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/UserCrudTest.java:38]: done

File: backend/src/test/java/com/google/sps/UserCrudTest.java (hunk 2 above)
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/UserCrudTest.java:32]: done

File: frontend/src/TripUtils.js (hunk 1 above)
Comment (RCAVelez) [reply frontend/src/TripUtils.js:5]: we decided to not return anything from the createUser, createTrip, and updateTrip servlets and utils

File: backend/src/test/java/com/google/sps/UserCrudTest.java [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+  @Test
+  public void readUserEmail() {
+    Entity userEntity = UserCrud.createUser(Email);
+    Entity readEntity = UserCrud.readEntity("email", Email, "User");
+    assertEquals(Email, readEntity.getProperty("email").toString());
+  }
+
+  @Test
+  public void noUserRead() {
```
[the remark this answers is not in the document]
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/UserCrudTest.java:71]: excellent! My tests are named better now!

File: backend/src/test/java/com/google/sps/UserCrudTest.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+public class UserCrudTest {
+  private static final LocalServiceTestHelper HELPER =
+      new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+  private static final String EMAIL = "testEMAIL@gmail.com";
+  private static final String EMAIL_2 = "testEMAIL_2@gmail.com";
+  private static final Long TRIP_ID = 111222333L;
+  private static final ArrayList<String> EMPTY_TRIP_IDS = new ArrayList<String>();
+  private static final ArrayList<String> SINGLE_TRIP_ID = new ArrayList<String>() {
+    { add(Long.toString(TRIP_ID)); }
+  };
```
Comment (mecolmg) [inline backend/src/test/java/com/google/sps/UserCrudTest.java:39]: nit: Import `java.util.Arrays` and you can do:
```suggestion
  private static final List<String> EMPTY_TRIP_IDS = new ArrayList<String>();
  private static final List<String> SINGLE_TRIP_ID = Arrays.asList({Long.toString(TRIP_ID)});
```

File: backend/src/test/java/com/google/sps/UserCrudTest.java [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+import java.util.ArrayList;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class UserCrudTest {
+  private static final LocalServiceTestHelper HELPER =
```
Comment (mecolmg) [inline backend/src/test/java/com/google/sps/UserCrudTest.java:31]: nit: Naming could be more descriptive (i.e. `MOCK_DATASTORE_SERVICE` or `DATASTORE_SERVICE_HELPER`).

File: backend/src/test/java/com/google/sps/UserCrudTest.java [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+    HELPER.setUp();
+  }
+
+  @After
+  public void tearDown() {
+    HELPER.tearDown();
+  }
+
+  @Test
+  public void createUser_returnsEmailForCreatedUser() {
```
Comment (mecolmg) [inline backend/src/test/java/com/google/sps/UserCrudTest.java:52]: Awesome test naming 👍

File: backend/src/test/java/com/google/sps/UserCrudTest.java [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+    Entity userEntity = UserCrud.createUser(EMAIL);
+    assertEquals(EMAIL, userEntity.getProperty("email").toString());
```
Comment (mecolmg) [inline backend/src/test/java/com/google/sps/UserCrudTest.java:54]: nit: I think the toString and casting are unnecessary, since assertEquals can compare two `Object`s

File: backend/src/test/java/com/google/sps/UserCrudTest.java [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+  @Test
+  public void readEntity_returnsEmailForFoundEntity() {
+    Entity userEntity = UserCrud.createUser(EMAIL);
+    Entity readEntity = UserCrud.readEntity("email", EMAIL, "User");
+    assertEquals(EMAIL, readEntity.getProperty("email").toString());
+  }
+
+  @Test
+  public void readEntity_returnsNullForMissingEntity() {
+    Entity userEntity = UserCrud.createUser(EMAIL);
```
Comment (mecolmg) [inline backend/src/test/java/com/google/sps/UserCrudTest.java:72]: nit: no need to create this user

File: backend/src/test/java/com/google/sps/UserCrudTest.java [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+
+  @Test
+  public void addTripId_returnsArrayListOfUpdateTripIds() {
+    Entity userEntity = UserCrud.createUser(EMAIL);
+    UserCrud.addTripId(EMAIL, TRIP_ID);
+    userEntity = UserCrud.readEntity("email", EMAIL, "User");
+    ArrayList<String> tripIds = (ArrayList<String>) userEntity.getProperty("tripIds");
+    assertEquals(SINGLE_TRIP_ID, tripIds);
+  }
+}
```
Comment (mecolmg) [inline backend/src/test/java/com/google/sps/UserCrudTest.java:84]: Please add test cases for:
- Calling createUser twice with the same email
- Calling addTripId twice with the same tripId
- Calling addTripId for an email that doesn't exist
- More edge cases if you can think of them. Look in your code for all those `if` statements and try to find lines that your tests don't cover.

File: backend/src/main/java/com/google/sps/UserCrud.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (genedwards) [reply backend/src/main/java/com/google/sps/UserCrud.java:30]: that's not my point though, on master, it already has `userEntity.setProperty("tripIds", new ArrayList<String>());`
https://github.com/googleinterns/step88-2020/blob/master/backend/src/main/java/com/google/sps/UserCrud.java#L30

which is a red flag that there's something going on with your branches

[the remark this answers is not in the document]
Comment (RCAVelez) [reply backend/src/main/java/com/google/sps/UserCrud.java:30]: Oh yes, this branch was out of sync with master as you mentioned. It should be sync now.

File: backend/src/test/java/com/google/sps/UserCrudTest.java (hunk 13 above)
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/UserCrudTest.java:31]: done

File: backend/src/test/java/com/google/sps/UserCrudTest.java (hunk 14 above)
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/UserCrudTest.java:52]: 😁

File: backend/src/test/java/com/google/sps/UserCrudTest.java (hunk 15 above)
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/UserCrudTest.java:54]: Wow, I learned something new! Yay!

File: backend/src/test/java/com/google/sps/UserCrudTest.java (hunk 16 above)
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/UserCrudTest.java:72]: done

File: backend/src/test/java/com/google/sps/UserCrudTest.java (hunk 12 above)
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/UserCrudTest.java:39]: done

File: backend/src/test/java/com/google/sps/UserCrudTest.java [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+  public void readEntity_returnsNullForMissingEntity() {
+    assertEquals(null, UserCrud.readEntity("email", EMAIL_2, "User"));
+  }
+
+  @Test
+  public void addTripId_returnsArrayListOfUpdateTripIds() {
+    Entity userEntity = UserCrud.createUser(EMAIL);
+    UserCrud.addTripId(EMAIL, TRIP_ID);
+    userEntity = UserCrud.readEntity("email", EMAIL, "User");
+    // ArrayList<String> tripIds = (ArrayList<String>) userEntity.getProperty("tripIds");
```
Comment (mecolmg) [inline backend/src/test/java/com/google/sps/UserCrudTest.java:80]: nit: Please remove.

Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/UserCrudTest.java:80]: done

File: backend/src/test/java/com/google/sps/UserCrudTest.java (hunk 17 above)
Comment (zoebarth) [reply backend/src/test/java/com/google/sps/UserCrudTest.java:102]: Done
